libs/core/kiln_ai/tools/test_mcp_session_manager.py

```diff
@@ -37,6 +47,176 @@ def test_singleton_reset_for_testing(self):
+        """Test _extract_first_exception with tuple of exception types."""
+        manager = MCPSessionManager()
+
+        # Test with matching type in tuple
+        error = ConnectionError("connection failed")
+        result = manager._extract_first_exception(error, (ValueError, ConnectionError))
+        assert result is error
+
+        # Test with non-matching types in tuple
+        result = manager._extract_first_exception(error, (TypeError, RuntimeError))
+        assert result is None
+
+    def test_extract_first_exception_with_exception_group(self):
+        """Test _extract_first_exception with ExceptionGroup (Python 3.11+ or backport)."""
+        manager = MCPSessionManager()
+
+        # Only run this test if ExceptionGroup is available
+        if "ExceptionGroup" not in globals():
+            pytest.skip("ExceptionGroup not available in this Python version")
+
+        # Create nested exceptions
+        inner_error = ValueError("inner error")
+        other_error = TypeError("other error")
+        group = ExceptionGroup("group error", [inner_error, other_error])
+
+        # Test extracting matching exception from group
+        result = manager._extract_first_exception(group, ValueError)
+        assert result is inner_error
+
+        # Test extracting non-matching exception from group
+        result = manager._extract_first_exception(group, RuntimeError)
+        assert result is None
+
+    def test_extract_first_exception_nested_exception_groups(self):
+        """Test _extract_first_exception with nested ExceptionGroups."""
+        manager = MCPSessionManager()
+
+        # Only run this test if ExceptionGroup is available
+        try:
+            # Check if ExceptionGroup is available
+            _ = ExceptionGroup
+        except NameError:
+            pytest.skip("ExceptionGroup not available in this Python version")
+
+        # Create deeply nested exception structure
+        target_error = FileNotFoundError("file not found")
+        inner_group = ExceptionGroup("inner group", [target_error, ValueError("other")])
+        outer_group = ExceptionGroup(
+            "outer group", [TypeError("type error"), inner_group]
+        )
+
+        # Should find the deeply nested exception
+        result = manager._extract_first_exception(outer_group, FileNotFoundError)
+        assert result is target_error
+
+        # Should find FileNotFoundError when searching for OSError (parent class)
+        result = manager._extract_first_exception(outer_group, OSError)
+        assert result is target_error
+
+        # Should not find non-existent exception type
+        result = manager._extract_first_exception(outer_group, KeyError)
+        assert result is None
+
+        # Should not find another unrelated exception type
+        result = manager._extract_first_exception(outer_group, LookupError)
+        assert result is None
+
+    def test_extract_first_exception_empty_exception_group(self):
```

Comment:
there's too much test code here for the use case. Did you read all of it and are signing up to maintain it all? No point in tests unless we're reviewed them and verify they all check the right things. Otherwise it's just maintenance cost. Sonnet will write tests that pass, but don't check important things or check incorrect things.

To eliminate repeated code, use pytest paramterize. Write the method once, and takes 2 parameters: the error to throw, and a string to check the resulting error matches. Should eliminate 90% of the code.